<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />

<meta name="author" content="Thang V Pham" />

<meta name="date" content="2024-12-03" />

<title>Fast MaxLFQ</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
</style>



<style type="text/css">
code {
white-space: pre;
}
.sourceCode {
overflow: visible;
}
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
{ counter-reset: source-line 0; }
pre.numberSource code > span
{ position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
{ content: counter(source-line);
position: relative; left: -1em; text-align: right; vertical-align: baseline;
border: none; display: inline-block;
-webkit-touch-callout: none; -webkit-user-select: none;
-khtml-user-select: none; -moz-user-select: none;
-ms-user-select: none; user-select: none;
padding: 0 4px; width: 4em;
color: #aaaaaa;
}
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa; padding-left: 4px; }
div.sourceCode
{ }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } 
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.at { color: #7d9029; } 
code span.bn { color: #40a070; } 
code span.bu { color: #008000; } 
code span.cf { color: #007020; font-weight: bold; } 
code span.ch { color: #4070a0; } 
code span.cn { color: #880000; } 
code span.co { color: #60a0b0; font-style: italic; } 
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.do { color: #ba2121; font-style: italic; } 
code span.dt { color: #902000; } 
code span.dv { color: #40a070; } 
code span.er { color: #ff0000; font-weight: bold; } 
code span.ex { } 
code span.fl { color: #40a070; } 
code span.fu { color: #06287e; } 
code span.im { color: #008000; font-weight: bold; } 
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.kw { color: #007020; font-weight: bold; } 
code span.op { color: #666666; } 
code span.ot { color: #007020; } 
code span.pp { color: #bc7a00; } 
code span.sc { color: #4070a0; } 
code span.ss { color: #bb6688; } 
code span.st { color: #4070a0; } 
code span.va { color: #19177c; } 
code span.vs { color: #4070a0; } 
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } 
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">Fast MaxLFQ</h1>
<h4 class="author">Thang V Pham</h4>
<h4 class="date">2024-12-03</h4>



<p>We have implemented a highly optimized version of the
<strong>iq</strong> pipeline (Pham et al., Bioinformatics 2020). To run
the following examples, download <a href="https://github.com/tvpham/iq/releases/download/v1.1/DIA-report-long-format.zip">DIA-report-long-format.zip</a>
and unzip the file to a local working directory.</p>
<p>The unzipped file <code>DIA-report-long-format.txt</code> is a
tab-deliminated text file exported from a Spectronaut search using this
export schema <a href="https://github.com/tvpham/iq/releases/download/v1.1/iq.rs">iq.rs</a>.
The user might want to import this schema to their Spectronaut
installation for the ease of using the <strong>iq</strong> pipeline.</p>
<div id="the-standard-pipeline" class="section level2">
<h2>The standard pipeline</h2>
<p>First, we apply the standard pipeline implemented in pure R. Read and
filter the data</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="fu">library</span>(<span class="st">&quot;iq&quot;</span>) <span class="co"># if not already installed, run install.packages(&quot;iq&quot;) </span></span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a></span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a>raw <span class="ot">&lt;-</span> <span class="fu">read.delim</span>(<span class="st">&quot;DIA-Report-long-format.txt&quot;</span>)</span>
<span id="cb1-4"><a href="#cb1-4" tabindex="-1"></a></span>
<span id="cb1-5"><a href="#cb1-5" tabindex="-1"></a>selected <span class="ot">&lt;-</span> raw<span class="sc">$</span>F.ExcludedFromQuantification <span class="sc">==</span> <span class="st">&quot;False&quot;</span> <span class="sc">&amp;</span> </span>
<span id="cb1-6"><a href="#cb1-6" tabindex="-1"></a>            <span class="sc">!</span><span class="fu">is.na</span>(raw<span class="sc">$</span>PG.Qvalue) <span class="sc">&amp;</span> (raw<span class="sc">$</span>PG.Qvalue <span class="sc">&lt;</span> <span class="fl">0.01</span>) <span class="sc">&amp;</span></span>
<span id="cb1-7"><a href="#cb1-7" tabindex="-1"></a>            <span class="sc">!</span><span class="fu">is.na</span>(raw<span class="sc">$</span>EG.Qvalue) <span class="sc">&amp;</span> (raw<span class="sc">$</span>EG.Qvalue <span class="sc">&lt;</span> <span class="fl">0.01</span>)</span>
<span id="cb1-8"><a href="#cb1-8" tabindex="-1"></a></span>
<span id="cb1-9"><a href="#cb1-9" tabindex="-1"></a>raw <span class="ot">&lt;-</span> raw[selected,]</span></code></pre></div>
<p>Normalize the data, create a protein list, and perform the MaxLFQ
algorithm</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a>sample_id  <span class="ot">&lt;-</span> <span class="st">&quot;R.FileName&quot;</span> </span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a></span>
<span id="cb2-3"><a href="#cb2-3" tabindex="-1"></a>secondary_id <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;EG.Library&quot;</span>, <span class="st">&quot;FG.Id&quot;</span>, <span class="st">&quot;FG.Charge&quot;</span>, <span class="st">&quot;F.FrgIon&quot;</span>, <span class="st">&quot;F.Charge&quot;</span>, <span class="st">&quot;F.FrgLossType&quot;</span>)</span>
<span id="cb2-4"><a href="#cb2-4" tabindex="-1"></a></span>
<span id="cb2-5"><a href="#cb2-5" tabindex="-1"></a>norm_data <span class="ot">&lt;-</span> iq<span class="sc">::</span><span class="fu">preprocess</span>(raw, </span>
<span id="cb2-6"><a href="#cb2-6" tabindex="-1"></a>                            <span class="at">sample_id  =</span> sample_id, </span>
<span id="cb2-7"><a href="#cb2-7" tabindex="-1"></a>                            <span class="at">secondary_id =</span> secondary_id)</span>
<span id="cb2-8"><a href="#cb2-8" tabindex="-1"></a><span class="co">#&gt; Concatenating secondary ids...</span></span>
<span id="cb2-9"><a href="#cb2-9" tabindex="-1"></a><span class="co">#&gt; Removing low intensities...</span></span>
<span id="cb2-10"><a href="#cb2-10" tabindex="-1"></a><span class="co">#&gt; Barplotting raw data ...</span></span>
<span id="cb2-11"><a href="#cb2-11" tabindex="-1"></a><span class="co">#&gt; Median normalization ...</span></span>
<span id="cb2-12"><a href="#cb2-12" tabindex="-1"></a><span class="co">#&gt; Barplotting after normalization ...</span></span>
<span id="cb2-13"><a href="#cb2-13" tabindex="-1"></a></span>
<span id="cb2-14"><a href="#cb2-14" tabindex="-1"></a>protein_list <span class="ot">&lt;-</span> iq<span class="sc">::</span><span class="fu">create_protein_list</span>(norm_data)</span>
<span id="cb2-15"><a href="#cb2-15" tabindex="-1"></a><span class="co">#&gt; # proteins = 3554, # samples = 24</span></span>
<span id="cb2-16"><a href="#cb2-16" tabindex="-1"></a><span class="co">#&gt; 5%</span></span>
<span id="cb2-17"><a href="#cb2-17" tabindex="-1"></a><span class="co">#&gt; 10%</span></span>
<span id="cb2-18"><a href="#cb2-18" tabindex="-1"></a><span class="co">#&gt; 15%</span></span>
<span id="cb2-19"><a href="#cb2-19" tabindex="-1"></a><span class="co">#&gt; 20%</span></span>
<span id="cb2-20"><a href="#cb2-20" tabindex="-1"></a><span class="co">#&gt; 25%</span></span>
<span id="cb2-21"><a href="#cb2-21" tabindex="-1"></a><span class="co">#&gt; 30%</span></span>
<span id="cb2-22"><a href="#cb2-22" tabindex="-1"></a><span class="co">#&gt; 35%</span></span>
<span id="cb2-23"><a href="#cb2-23" tabindex="-1"></a><span class="co">#&gt; 40%</span></span>
<span id="cb2-24"><a href="#cb2-24" tabindex="-1"></a><span class="co">#&gt; 45%</span></span>
<span id="cb2-25"><a href="#cb2-25" tabindex="-1"></a><span class="co">#&gt; 50%</span></span>
<span id="cb2-26"><a href="#cb2-26" tabindex="-1"></a><span class="co">#&gt; 55%</span></span>
<span id="cb2-27"><a href="#cb2-27" tabindex="-1"></a><span class="co">#&gt; 60%</span></span>
<span id="cb2-28"><a href="#cb2-28" tabindex="-1"></a><span class="co">#&gt; 65%</span></span>
<span id="cb2-29"><a href="#cb2-29" tabindex="-1"></a><span class="co">#&gt; 70%</span></span>
<span id="cb2-30"><a href="#cb2-30" tabindex="-1"></a><span class="co">#&gt; 75%</span></span>
<span id="cb2-31"><a href="#cb2-31" tabindex="-1"></a><span class="co">#&gt; 80%</span></span>
<span id="cb2-32"><a href="#cb2-32" tabindex="-1"></a><span class="co">#&gt; 85%</span></span>
<span id="cb2-33"><a href="#cb2-33" tabindex="-1"></a><span class="co">#&gt; 90%</span></span>
<span id="cb2-34"><a href="#cb2-34" tabindex="-1"></a><span class="co">#&gt; 95%</span></span>
<span id="cb2-35"><a href="#cb2-35" tabindex="-1"></a><span class="co">#&gt; 100%</span></span>
<span id="cb2-36"><a href="#cb2-36" tabindex="-1"></a><span class="co">#&gt; Completed.</span></span>
<span id="cb2-37"><a href="#cb2-37" tabindex="-1"></a></span>
<span id="cb2-38"><a href="#cb2-38" tabindex="-1"></a>result <span class="ot">&lt;-</span> iq<span class="sc">::</span><span class="fu">create_protein_table</span>(protein_list)</span>
<span id="cb2-39"><a href="#cb2-39" tabindex="-1"></a><span class="co">#&gt; 5%</span></span>
<span id="cb2-40"><a href="#cb2-40" tabindex="-1"></a><span class="co">#&gt; 10%</span></span>
<span id="cb2-41"><a href="#cb2-41" tabindex="-1"></a><span class="co">#&gt; 15%</span></span>
<span id="cb2-42"><a href="#cb2-42" tabindex="-1"></a><span class="co">#&gt; 20%</span></span>
<span id="cb2-43"><a href="#cb2-43" tabindex="-1"></a><span class="co">#&gt; 25%</span></span>
<span id="cb2-44"><a href="#cb2-44" tabindex="-1"></a><span class="co">#&gt; 30%</span></span>
<span id="cb2-45"><a href="#cb2-45" tabindex="-1"></a><span class="co">#&gt; 35%</span></span>
<span id="cb2-46"><a href="#cb2-46" tabindex="-1"></a><span class="co">#&gt; 40%</span></span>
<span id="cb2-47"><a href="#cb2-47" tabindex="-1"></a><span class="co">#&gt; 45%</span></span>
<span id="cb2-48"><a href="#cb2-48" tabindex="-1"></a><span class="co">#&gt; 50%</span></span>
<span id="cb2-49"><a href="#cb2-49" tabindex="-1"></a><span class="co">#&gt; 55%</span></span>
<span id="cb2-50"><a href="#cb2-50" tabindex="-1"></a><span class="co">#&gt; 60%</span></span>
<span id="cb2-51"><a href="#cb2-51" tabindex="-1"></a><span class="co">#&gt; 65%</span></span>
<span id="cb2-52"><a href="#cb2-52" tabindex="-1"></a><span class="co">#&gt; 70%</span></span>
<span id="cb2-53"><a href="#cb2-53" tabindex="-1"></a><span class="co">#&gt; 75%</span></span>
<span id="cb2-54"><a href="#cb2-54" tabindex="-1"></a><span class="co">#&gt; 80%</span></span>
<span id="cb2-55"><a href="#cb2-55" tabindex="-1"></a><span class="co">#&gt; 85%</span></span>
<span id="cb2-56"><a href="#cb2-56" tabindex="-1"></a><span class="co">#&gt; 90%</span></span>
<span id="cb2-57"><a href="#cb2-57" tabindex="-1"></a><span class="co">#&gt; 95%</span></span>
<span id="cb2-58"><a href="#cb2-58" tabindex="-1"></a><span class="co">#&gt; 100%</span></span>
<span id="cb2-59"><a href="#cb2-59" tabindex="-1"></a><span class="co">#&gt; Completed.</span></span></code></pre></div>
<p>Extract annotation columns and write the result to an output file</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a>annotation_columns <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;PG.Genes&quot;</span>, <span class="st">&quot;PG.ProteinNames&quot;</span>)</span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a></span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a>extra_names <span class="ot">&lt;-</span> iq<span class="sc">::</span><span class="fu">extract_annotation</span>(<span class="fu">rownames</span>(result<span class="sc">$</span>estimate), </span>
<span id="cb3-4"><a href="#cb3-4" tabindex="-1"></a>                                      raw, </span>
<span id="cb3-5"><a href="#cb3-5" tabindex="-1"></a>                                      <span class="at">annotation_columns =</span> annotation_columns)</span>
<span id="cb3-6"><a href="#cb3-6" tabindex="-1"></a></span>
<span id="cb3-7"><a href="#cb3-7" tabindex="-1"></a><span class="fu">write.table</span>(<span class="fu">cbind</span>(<span class="at">Protein =</span> <span class="fu">rownames</span>(result<span class="sc">$</span>estimate),</span>
<span id="cb3-8"><a href="#cb3-8" tabindex="-1"></a>                  extra_names[, annotation_columns],</span>
<span id="cb3-9"><a href="#cb3-9" tabindex="-1"></a>                  <span class="at">MaxLFQ_annotation =</span> result<span class="sc">$</span>annotation,</span>
<span id="cb3-10"><a href="#cb3-10" tabindex="-1"></a>                  result<span class="sc">$</span>estimate),</span>
<span id="cb3-11"><a href="#cb3-11" tabindex="-1"></a>            <span class="st">&quot;iq-MaxLFQ.txt&quot;</span>, <span class="at">sep =</span> <span class="st">&quot;</span><span class="sc">\t</span><span class="st">&quot;</span>, <span class="at">row.names =</span> <span class="cn">FALSE</span>)</span></code></pre></div>
<p>The resulting file <code>iq-MaxLFQ.txt</code> is the protein level
quantification report in a tab-deliminated text format.</p>
</div>
<div id="a-faster-maxlfq-implementation" class="section level2">
<h2>A faster MaxLFQ implementation</h2>
<p>The function <code>iq::fast_MaxLFQ</code> implemented in C++ combines
the functionalities of <code>iq::create_protein_list</code> and
<code>iq::create_protein_table</code>.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="co">#--------------------- Replacing ---------------------</span></span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a><span class="co"># protein_list &lt;- iq::create_protein_list(norm_data) #</span></span>
<span id="cb4-3"><a href="#cb4-3" tabindex="-1"></a><span class="co"># result &lt;- iq::create_protein_table(protein_list)   #</span></span>
<span id="cb4-4"><a href="#cb4-4" tabindex="-1"></a><span class="co">#-----------------------------------------------------</span></span>
<span id="cb4-5"><a href="#cb4-5" tabindex="-1"></a></span>
<span id="cb4-6"><a href="#cb4-6" tabindex="-1"></a>result_faster <span class="ot">&lt;-</span> iq<span class="sc">::</span><span class="fu">fast_MaxLFQ</span>(norm_data)</span>
<span id="cb4-7"><a href="#cb4-7" tabindex="-1"></a><span class="co">#&gt; nrow = 3369557, # proteins = 3554, # samples = 24</span></span>
<span id="cb4-8"><a href="#cb4-8" tabindex="-1"></a><span class="co">#&gt; Using 35 threads...</span></span>
<span id="cb4-9"><a href="#cb4-9" tabindex="-1"></a><span class="co">#&gt; 0%</span></span>
<span id="cb4-10"><a href="#cb4-10" tabindex="-1"></a><span class="co">#&gt; 6%</span></span>
<span id="cb4-11"><a href="#cb4-11" tabindex="-1"></a><span class="co">#&gt; 13%</span></span>
<span id="cb4-12"><a href="#cb4-12" tabindex="-1"></a><span class="co">#&gt; 18%</span></span>
<span id="cb4-13"><a href="#cb4-13" tabindex="-1"></a><span class="co">#&gt; 24%</span></span>
<span id="cb4-14"><a href="#cb4-14" tabindex="-1"></a><span class="co">#&gt; 29%</span></span>
<span id="cb4-15"><a href="#cb4-15" tabindex="-1"></a><span class="co">#&gt; 36%</span></span>
<span id="cb4-16"><a href="#cb4-16" tabindex="-1"></a><span class="co">#&gt; 42%</span></span>
<span id="cb4-17"><a href="#cb4-17" tabindex="-1"></a><span class="co">#&gt; 47%</span></span>
<span id="cb4-18"><a href="#cb4-18" tabindex="-1"></a><span class="co">#&gt; 52%</span></span>
<span id="cb4-19"><a href="#cb4-19" tabindex="-1"></a><span class="co">#&gt; 57%</span></span>
<span id="cb4-20"><a href="#cb4-20" tabindex="-1"></a><span class="co">#&gt; 64%</span></span>
<span id="cb4-21"><a href="#cb4-21" tabindex="-1"></a><span class="co">#&gt; 70%</span></span>
<span id="cb4-22"><a href="#cb4-22" tabindex="-1"></a><span class="co">#&gt; 77%</span></span>
<span id="cb4-23"><a href="#cb4-23" tabindex="-1"></a><span class="co">#&gt; 82%</span></span>
<span id="cb4-24"><a href="#cb4-24" tabindex="-1"></a><span class="co">#&gt; 87%</span></span>
<span id="cb4-25"><a href="#cb4-25" tabindex="-1"></a><span class="co">#&gt; 92%</span></span>
<span id="cb4-26"><a href="#cb4-26" tabindex="-1"></a><span class="co">#&gt; 98%</span></span>
<span id="cb4-27"><a href="#cb4-27" tabindex="-1"></a><span class="co">#&gt; Completed.</span></span></code></pre></div>
<p>The results of the R implementation <code>result</code> and C++
implementation <code>result_faster</code> should be equal up to the
floating-point precision of the underlying numerical libraries.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">&quot;Max difference =&quot;</span>, <span class="fu">max</span>(<span class="fu">abs</span>(result_faster<span class="sc">$</span>estimate <span class="sc">-</span> result<span class="sc">$</span>estimate), <span class="at">na.rm =</span> <span class="cn">TRUE</span>), <span class="st">&quot;</span><span class="sc">\n</span><span class="st">&quot;</span>)</span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a><span class="co">#&gt; Max difference = 1.332268e-13</span></span>
<span id="cb5-3"><a href="#cb5-3" tabindex="-1"></a></span>
<span id="cb5-4"><a href="#cb5-4" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">&quot;Identical NAs =&quot;</span>, <span class="fu">identical</span>(<span class="fu">is.na</span>(result_faster<span class="sc">$</span>estimate), <span class="fu">is.na</span>(result<span class="sc">$</span>estimate)), <span class="st">&quot;</span><span class="sc">\n</span><span class="st">&quot;</span>)</span>
<span id="cb5-5"><a href="#cb5-5" tabindex="-1"></a><span class="co">#&gt; Identical NAs = TRUE</span></span>
<span id="cb5-6"><a href="#cb5-6" tabindex="-1"></a></span>
<span id="cb5-7"><a href="#cb5-7" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">&quot;Equal annotation =&quot;</span>, <span class="fu">identical</span>(result_faster<span class="sc">$</span>annotation, result<span class="sc">$</span>annotation), <span class="st">&quot;</span><span class="sc">\n</span><span class="st">&quot;</span>)</span>
<span id="cb5-8"><a href="#cb5-8" tabindex="-1"></a><span class="co">#&gt; Equal annotation = TRUE</span></span></code></pre></div>
<div id="benchmarking-execution-time" class="section level3">
<h3>Benchmarking execution time</h3>
<p>We can check the improvement in execution time. The following result
is obtained on a computer with 12 CPU cores.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="fu">system.time</span>({</span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a>    protein_list <span class="ot">&lt;-</span> iq<span class="sc">::</span><span class="fu">create_protein_list</span>(norm_data)</span>
<span id="cb6-3"><a href="#cb6-3" tabindex="-1"></a>    result <span class="ot">&lt;-</span> iq<span class="sc">::</span><span class="fu">create_protein_table</span>(protein_list)</span>
<span id="cb6-4"><a href="#cb6-4" tabindex="-1"></a>})</span>
<span id="cb6-5"><a href="#cb6-5" tabindex="-1"></a><span class="co">#&gt; # proteins = 3554, # samples = 24</span></span>
<span id="cb6-6"><a href="#cb6-6" tabindex="-1"></a><span class="co">#&gt; 5%</span></span>
<span id="cb6-7"><a href="#cb6-7" tabindex="-1"></a><span class="co">#&gt; 10%</span></span>
<span id="cb6-8"><a href="#cb6-8" tabindex="-1"></a><span class="co">#&gt; 15%</span></span>
<span id="cb6-9"><a href="#cb6-9" tabindex="-1"></a><span class="co">#&gt; 20%</span></span>
<span id="cb6-10"><a href="#cb6-10" tabindex="-1"></a><span class="co">#&gt; 25%</span></span>
<span id="cb6-11"><a href="#cb6-11" tabindex="-1"></a><span class="co">#&gt; 30%</span></span>
<span id="cb6-12"><a href="#cb6-12" tabindex="-1"></a><span class="co">#&gt; 35%</span></span>
<span id="cb6-13"><a href="#cb6-13" tabindex="-1"></a><span class="co">#&gt; 40%</span></span>
<span id="cb6-14"><a href="#cb6-14" tabindex="-1"></a><span class="co">#&gt; 45%</span></span>
<span id="cb6-15"><a href="#cb6-15" tabindex="-1"></a><span class="co">#&gt; 50%</span></span>
<span id="cb6-16"><a href="#cb6-16" tabindex="-1"></a><span class="co">#&gt; 55%</span></span>
<span id="cb6-17"><a href="#cb6-17" tabindex="-1"></a><span class="co">#&gt; 60%</span></span>
<span id="cb6-18"><a href="#cb6-18" tabindex="-1"></a><span class="co">#&gt; 65%</span></span>
<span id="cb6-19"><a href="#cb6-19" tabindex="-1"></a><span class="co">#&gt; 70%</span></span>
<span id="cb6-20"><a href="#cb6-20" tabindex="-1"></a><span class="co">#&gt; 75%</span></span>
<span id="cb6-21"><a href="#cb6-21" tabindex="-1"></a><span class="co">#&gt; 80%</span></span>
<span id="cb6-22"><a href="#cb6-22" tabindex="-1"></a><span class="co">#&gt; 85%</span></span>
<span id="cb6-23"><a href="#cb6-23" tabindex="-1"></a><span class="co">#&gt; 90%</span></span>
<span id="cb6-24"><a href="#cb6-24" tabindex="-1"></a><span class="co">#&gt; 95%</span></span>
<span id="cb6-25"><a href="#cb6-25" tabindex="-1"></a><span class="co">#&gt; 100%</span></span>
<span id="cb6-26"><a href="#cb6-26" tabindex="-1"></a><span class="co">#&gt; Completed.</span></span>
<span id="cb6-27"><a href="#cb6-27" tabindex="-1"></a><span class="co">#&gt; 5%</span></span>
<span id="cb6-28"><a href="#cb6-28" tabindex="-1"></a><span class="co">#&gt; 10%</span></span>
<span id="cb6-29"><a href="#cb6-29" tabindex="-1"></a><span class="co">#&gt; 15%</span></span>
<span id="cb6-30"><a href="#cb6-30" tabindex="-1"></a><span class="co">#&gt; 20%</span></span>
<span id="cb6-31"><a href="#cb6-31" tabindex="-1"></a><span class="co">#&gt; 25%</span></span>
<span id="cb6-32"><a href="#cb6-32" tabindex="-1"></a><span class="co">#&gt; 30%</span></span>
<span id="cb6-33"><a href="#cb6-33" tabindex="-1"></a><span class="co">#&gt; 35%</span></span>
<span id="cb6-34"><a href="#cb6-34" tabindex="-1"></a><span class="co">#&gt; 40%</span></span>
<span id="cb6-35"><a href="#cb6-35" tabindex="-1"></a><span class="co">#&gt; 45%</span></span>
<span id="cb6-36"><a href="#cb6-36" tabindex="-1"></a><span class="co">#&gt; 50%</span></span>
<span id="cb6-37"><a href="#cb6-37" tabindex="-1"></a><span class="co">#&gt; 55%</span></span>
<span id="cb6-38"><a href="#cb6-38" tabindex="-1"></a><span class="co">#&gt; 60%</span></span>
<span id="cb6-39"><a href="#cb6-39" tabindex="-1"></a><span class="co">#&gt; 65%</span></span>
<span id="cb6-40"><a href="#cb6-40" tabindex="-1"></a><span class="co">#&gt; 70%</span></span>
<span id="cb6-41"><a href="#cb6-41" tabindex="-1"></a><span class="co">#&gt; 75%</span></span>
<span id="cb6-42"><a href="#cb6-42" tabindex="-1"></a><span class="co">#&gt; 80%</span></span>
<span id="cb6-43"><a href="#cb6-43" tabindex="-1"></a><span class="co">#&gt; 85%</span></span>
<span id="cb6-44"><a href="#cb6-44" tabindex="-1"></a><span class="co">#&gt; 90%</span></span>
<span id="cb6-45"><a href="#cb6-45" tabindex="-1"></a><span class="co">#&gt; 95%</span></span>
<span id="cb6-46"><a href="#cb6-46" tabindex="-1"></a><span class="co">#&gt; 100%</span></span>
<span id="cb6-47"><a href="#cb6-47" tabindex="-1"></a><span class="co">#&gt; Completed.</span></span>
<span id="cb6-48"><a href="#cb6-48" tabindex="-1"></a><span class="co">#&gt;    user  system elapsed </span></span>
<span id="cb6-49"><a href="#cb6-49" tabindex="-1"></a><span class="co">#&gt;  395.82    9.00  404.87</span></span>
<span id="cb6-50"><a href="#cb6-50" tabindex="-1"></a></span>
<span id="cb6-51"><a href="#cb6-51" tabindex="-1"></a><span class="fu">system.time</span>({</span>
<span id="cb6-52"><a href="#cb6-52" tabindex="-1"></a>    result_faster <span class="ot">&lt;-</span> iq<span class="sc">::</span><span class="fu">fast_MaxLFQ</span>(norm_data)</span>
<span id="cb6-53"><a href="#cb6-53" tabindex="-1"></a>})</span>
<span id="cb6-54"><a href="#cb6-54" tabindex="-1"></a><span class="co">#&gt; nrow = 3369557, # proteins = 3554, # samples = 24</span></span>
<span id="cb6-55"><a href="#cb6-55" tabindex="-1"></a><span class="co">#&gt; Using 35 threads...</span></span>
<span id="cb6-56"><a href="#cb6-56" tabindex="-1"></a><span class="co">#&gt; 0%</span></span>
<span id="cb6-57"><a href="#cb6-57" tabindex="-1"></a><span class="co">#&gt; 7%</span></span>
<span id="cb6-58"><a href="#cb6-58" tabindex="-1"></a><span class="co">#&gt; 15%</span></span>
<span id="cb6-59"><a href="#cb6-59" tabindex="-1"></a><span class="co">#&gt; 24%</span></span>
<span id="cb6-60"><a href="#cb6-60" tabindex="-1"></a><span class="co">#&gt; 30%</span></span>
<span id="cb6-61"><a href="#cb6-61" tabindex="-1"></a><span class="co">#&gt; 36%</span></span>
<span id="cb6-62"><a href="#cb6-62" tabindex="-1"></a><span class="co">#&gt; 42%</span></span>
<span id="cb6-63"><a href="#cb6-63" tabindex="-1"></a><span class="co">#&gt; 47%</span></span>
<span id="cb6-64"><a href="#cb6-64" tabindex="-1"></a><span class="co">#&gt; 54%</span></span>
<span id="cb6-65"><a href="#cb6-65" tabindex="-1"></a><span class="co">#&gt; 59%</span></span>
<span id="cb6-66"><a href="#cb6-66" tabindex="-1"></a><span class="co">#&gt; 65%</span></span>
<span id="cb6-67"><a href="#cb6-67" tabindex="-1"></a><span class="co">#&gt; 72%</span></span>
<span id="cb6-68"><a href="#cb6-68" tabindex="-1"></a><span class="co">#&gt; 77%</span></span>
<span id="cb6-69"><a href="#cb6-69" tabindex="-1"></a><span class="co">#&gt; 83%</span></span>
<span id="cb6-70"><a href="#cb6-70" tabindex="-1"></a><span class="co">#&gt; 90%</span></span>
<span id="cb6-71"><a href="#cb6-71" tabindex="-1"></a><span class="co">#&gt; 96%</span></span>
<span id="cb6-72"><a href="#cb6-72" tabindex="-1"></a><span class="co">#&gt; Completed.</span></span>
<span id="cb6-73"><a href="#cb6-73" tabindex="-1"></a><span class="co">#&gt;    user  system elapsed </span></span>
<span id="cb6-74"><a href="#cb6-74" tabindex="-1"></a><span class="co">#&gt;    7.75    0.04    3.55</span></span></code></pre></div>
</div>
</div>
<div id="an-efficient-data-structure-and-data-loading" class="section level2">
<h2>An efficient data structure and data loading</h2>
<p>We have implemented a fast data loading algorithm and an efficient
data structure. The memory usage is highly optimized to enable the
processing of very large datasets.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a>sample_id  <span class="ot">&lt;-</span> <span class="st">&quot;R.FileName&quot;</span> </span>
<span id="cb7-2"><a href="#cb7-2" tabindex="-1"></a></span>
<span id="cb7-3"><a href="#cb7-3" tabindex="-1"></a>secondary_id <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;EG.Library&quot;</span>, <span class="st">&quot;FG.Id&quot;</span>, <span class="st">&quot;FG.Charge&quot;</span>, <span class="st">&quot;F.FrgIon&quot;</span>, <span class="st">&quot;F.Charge&quot;</span>, <span class="st">&quot;F.FrgLossType&quot;</span>)</span>
<span id="cb7-4"><a href="#cb7-4" tabindex="-1"></a></span>
<span id="cb7-5"><a href="#cb7-5" tabindex="-1"></a>annotation_columns <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;PG.Genes&quot;</span>, <span class="st">&quot;PG.ProteinNames&quot;</span>)</span>
<span id="cb7-6"><a href="#cb7-6" tabindex="-1"></a></span>
<span id="cb7-7"><a href="#cb7-7" tabindex="-1"></a>iq_dat <span class="ot">&lt;-</span> iq<span class="sc">::</span><span class="fu">fast_read</span>(<span class="st">&quot;DIA-report-long-format.txt&quot;</span>,</span>
<span id="cb7-8"><a href="#cb7-8" tabindex="-1"></a>                        <span class="at">sample_id  =</span> sample_id, </span>
<span id="cb7-9"><a href="#cb7-9" tabindex="-1"></a>                        <span class="at">secondary_id =</span> secondary_id,</span>
<span id="cb7-10"><a href="#cb7-10" tabindex="-1"></a>                        <span class="at">filter_string_equal =</span> <span class="fu">c</span>(<span class="st">&quot;F.ExcludedFromQuantification&quot;</span> <span class="ot">=</span> <span class="st">&quot;False&quot;</span>),</span>
<span id="cb7-11"><a href="#cb7-11" tabindex="-1"></a>                        <span class="at">annotation_col =</span> annotation_columns)</span>
<span id="cb7-12"><a href="#cb7-12" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb7-13"><a href="#cb7-13" tabindex="-1"></a><span class="co">#&gt; Command: --sample R.FileName --primary PG.ProteinGroups --secondary EG.Library FG.Id FG.Charge F.FrgIon F.Charge F.FrgLossType --quant F.PeakArea --annotation PG.Genes PG.ProteinNames --filter-string-equal F.ExcludedFromQuantification False --filter-double-less PG.Qvalue 0.01 --filter-double-less EG.Qvalue 0.01 DIA-report-long-format.txt </span></span>
<span id="cb7-14"><a href="#cb7-14" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb7-15"><a href="#cb7-15" tabindex="-1"></a><span class="co">#&gt; Sample column:</span></span>
<span id="cb7-16"><a href="#cb7-16" tabindex="-1"></a><span class="co">#&gt;     R.FileName</span></span>
<span id="cb7-17"><a href="#cb7-17" tabindex="-1"></a><span class="co">#&gt; Protein column:</span></span>
<span id="cb7-18"><a href="#cb7-18" tabindex="-1"></a><span class="co">#&gt;     PG.ProteinGroups</span></span>
<span id="cb7-19"><a href="#cb7-19" tabindex="-1"></a><span class="co">#&gt; Ion column(s):</span></span>
<span id="cb7-20"><a href="#cb7-20" tabindex="-1"></a><span class="co">#&gt;     EG.Library FG.Id FG.Charge F.FrgIon F.Charge F.FrgLossType</span></span>
<span id="cb7-21"><a href="#cb7-21" tabindex="-1"></a><span class="co">#&gt; Quant column:</span></span>
<span id="cb7-22"><a href="#cb7-22" tabindex="-1"></a><span class="co">#&gt;     F.PeakArea</span></span>
<span id="cb7-23"><a href="#cb7-23" tabindex="-1"></a><span class="co">#&gt; Annotation column(s):</span></span>
<span id="cb7-24"><a href="#cb7-24" tabindex="-1"></a><span class="co">#&gt;     PG.Genes PG.ProteinNames</span></span>
<span id="cb7-25"><a href="#cb7-25" tabindex="-1"></a><span class="co">#&gt; String equal filter(s):</span></span>
<span id="cb7-26"><a href="#cb7-26" tabindex="-1"></a><span class="co">#&gt;     F.ExcludedFromQuantification == False</span></span>
<span id="cb7-27"><a href="#cb7-27" tabindex="-1"></a><span class="co">#&gt; Double less filter(s):</span></span>
<span id="cb7-28"><a href="#cb7-28" tabindex="-1"></a><span class="co">#&gt;     PG.Qvalue &lt; 0.010000</span></span>
<span id="cb7-29"><a href="#cb7-29" tabindex="-1"></a><span class="co">#&gt;     EG.Qvalue &lt; 0.010000</span></span>
<span id="cb7-30"><a href="#cb7-30" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb7-31"><a href="#cb7-31" tabindex="-1"></a><span class="co">#&gt; Using 4 threads ...</span></span>
<span id="cb7-32"><a href="#cb7-32" tabindex="-1"></a><span class="co">#&gt; 20 samples read</span></span>
<span id="cb7-33"><a href="#cb7-33" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb7-34"><a href="#cb7-34" tabindex="-1"></a><span class="co">#&gt; # lines read (excluding headers)      = 5547331</span></span>
<span id="cb7-35"><a href="#cb7-35" tabindex="-1"></a><span class="co">#&gt; # quantitative values after filtering = 3390569</span></span>
<span id="cb7-36"><a href="#cb7-36" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb7-37"><a href="#cb7-37" tabindex="-1"></a><span class="co">#&gt; # samples  = 24</span></span>
<span id="cb7-38"><a href="#cb7-38" tabindex="-1"></a><span class="co">#&gt; # proteins = 3554</span></span>
<span id="cb7-39"><a href="#cb7-39" tabindex="-1"></a></span>
<span id="cb7-40"><a href="#cb7-40" tabindex="-1"></a>iq_norm_data <span class="ot">&lt;-</span> iq<span class="sc">::</span><span class="fu">fast_preprocess</span>(iq_dat<span class="sc">$</span>quant_table)</span>
<span id="cb7-41"><a href="#cb7-41" tabindex="-1"></a><span class="co">#&gt; Removing low intensities...</span></span>
<span id="cb7-42"><a href="#cb7-42" tabindex="-1"></a><span class="co">#&gt; Barplotting raw data ...</span></span>
<span id="cb7-43"><a href="#cb7-43" tabindex="-1"></a><span class="co">#&gt; Median normalization ...</span></span>
<span id="cb7-44"><a href="#cb7-44" tabindex="-1"></a><span class="co">#&gt; Barplotting after normalization ...</span></span>
<span id="cb7-45"><a href="#cb7-45" tabindex="-1"></a></span>
<span id="cb7-46"><a href="#cb7-46" tabindex="-1"></a>result_fastest <span class="ot">&lt;-</span> iq<span class="sc">::</span><span class="fu">fast_MaxLFQ</span>(iq_norm_data, </span>
<span id="cb7-47"><a href="#cb7-47" tabindex="-1"></a>                                  <span class="at">row_names =</span> iq_dat<span class="sc">$</span>protein[, <span class="dv">1</span>], </span>
<span id="cb7-48"><a href="#cb7-48" tabindex="-1"></a>                                  <span class="at">col_names =</span> iq_dat<span class="sc">$</span>sample)</span>
<span id="cb7-49"><a href="#cb7-49" tabindex="-1"></a><span class="co">#&gt; nrow = 3369557, # proteins = 3554, # samples = 24</span></span>
<span id="cb7-50"><a href="#cb7-50" tabindex="-1"></a><span class="co">#&gt; Using 35 threads...</span></span>
<span id="cb7-51"><a href="#cb7-51" tabindex="-1"></a><span class="co">#&gt; 0%</span></span>
<span id="cb7-52"><a href="#cb7-52" tabindex="-1"></a><span class="co">#&gt; 5%</span></span>
<span id="cb7-53"><a href="#cb7-53" tabindex="-1"></a><span class="co">#&gt; 11%</span></span>
<span id="cb7-54"><a href="#cb7-54" tabindex="-1"></a><span class="co">#&gt; 16%</span></span>
<span id="cb7-55"><a href="#cb7-55" tabindex="-1"></a><span class="co">#&gt; 22%</span></span>
<span id="cb7-56"><a href="#cb7-56" tabindex="-1"></a><span class="co">#&gt; 27%</span></span>
<span id="cb7-57"><a href="#cb7-57" tabindex="-1"></a><span class="co">#&gt; 33%</span></span>
<span id="cb7-58"><a href="#cb7-58" tabindex="-1"></a><span class="co">#&gt; 40%</span></span>
<span id="cb7-59"><a href="#cb7-59" tabindex="-1"></a><span class="co">#&gt; 46%</span></span>
<span id="cb7-60"><a href="#cb7-60" tabindex="-1"></a><span class="co">#&gt; 53%</span></span>
<span id="cb7-61"><a href="#cb7-61" tabindex="-1"></a><span class="co">#&gt; 58%</span></span>
<span id="cb7-62"><a href="#cb7-62" tabindex="-1"></a><span class="co">#&gt; 63%</span></span>
<span id="cb7-63"><a href="#cb7-63" tabindex="-1"></a><span class="co">#&gt; 69%</span></span>
<span id="cb7-64"><a href="#cb7-64" tabindex="-1"></a><span class="co">#&gt; 87%</span></span>
<span id="cb7-65"><a href="#cb7-65" tabindex="-1"></a><span class="co">#&gt; 93%</span></span>
<span id="cb7-66"><a href="#cb7-66" tabindex="-1"></a><span class="co">#&gt; 99%</span></span>
<span id="cb7-67"><a href="#cb7-67" tabindex="-1"></a><span class="co">#&gt; Completed.</span></span></code></pre></div>
<p>The result of the optimized pipeline <code>result_fastest</code>
should be the same as that of the standard pipeline
<code>result</code>.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">&quot;Max difference =&quot;</span>, <span class="fu">max</span>(<span class="fu">abs</span>(result_fastest<span class="sc">$</span>estimate <span class="sc">-</span> result<span class="sc">$</span>estimate), <span class="at">na.rm =</span> <span class="cn">TRUE</span>), <span class="st">&quot;</span><span class="sc">\n</span><span class="st">&quot;</span>)</span>
<span id="cb8-2"><a href="#cb8-2" tabindex="-1"></a><span class="co">#&gt; Max difference = 1.136868e-13</span></span>
<span id="cb8-3"><a href="#cb8-3" tabindex="-1"></a></span>
<span id="cb8-4"><a href="#cb8-4" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">&quot;Identical NAs =&quot;</span>, <span class="fu">identical</span>(<span class="fu">is.na</span>(result_fastest<span class="sc">$</span>estimate), <span class="fu">is.na</span>(result<span class="sc">$</span>estimate)), <span class="st">&quot;</span><span class="sc">\n</span><span class="st">&quot;</span>)</span>
<span id="cb8-5"><a href="#cb8-5" tabindex="-1"></a><span class="co">#&gt; Identical NAs = TRUE</span></span>
<span id="cb8-6"><a href="#cb8-6" tabindex="-1"></a></span>
<span id="cb8-7"><a href="#cb8-7" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">&quot;Equal annotation =&quot;</span>, <span class="fu">identical</span>(result_fastest<span class="sc">$</span>annotation, result<span class="sc">$</span>annotation), <span class="st">&quot;</span><span class="sc">\n</span><span class="st">&quot;</span>)</span>
<span id="cb8-8"><a href="#cb8-8" tabindex="-1"></a><span class="co">#&gt; Equal annotation = TRUE</span></span></code></pre></div>
<p>The annotation columns are stored in the <code>protein</code>
component of the input data structure <code>iq_dat</code>. We can
extract the annotation columns and write the result to an output text
file.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a>iq_extra_names <span class="ot">&lt;-</span> iq<span class="sc">::</span><span class="fu">extract_annotation</span>(<span class="fu">rownames</span>(result_fastest<span class="sc">$</span>estimate), </span>
<span id="cb9-2"><a href="#cb9-2" tabindex="-1"></a>                                         iq_dat<span class="sc">$</span>protein, </span>
<span id="cb9-3"><a href="#cb9-3" tabindex="-1"></a>                                         <span class="at">annotation_columns =</span> annotation_columns)</span>
<span id="cb9-4"><a href="#cb9-4" tabindex="-1"></a></span>
<span id="cb9-5"><a href="#cb9-5" tabindex="-1"></a><span class="fu">write.table</span>(<span class="fu">cbind</span>(<span class="at">Protein =</span> <span class="fu">rownames</span>(result_fastest<span class="sc">$</span>estimate),</span>
<span id="cb9-6"><a href="#cb9-6" tabindex="-1"></a>                  iq_extra_names[, annotation_columns],</span>
<span id="cb9-7"><a href="#cb9-7" tabindex="-1"></a>                  <span class="at">MaxLFQ_annotation =</span> result_fastest<span class="sc">$</span>annotation,</span>
<span id="cb9-8"><a href="#cb9-8" tabindex="-1"></a>                  result_fastest<span class="sc">$</span>estimate), </span>
<span id="cb9-9"><a href="#cb9-9" tabindex="-1"></a>            <span class="st">&quot;iq-MaxLFQ-fast.txt&quot;</span>, <span class="at">sep =</span> <span class="st">&quot;</span><span class="sc">\t</span><span class="st">&quot;</span>, <span class="at">row.names =</span> <span class="cn">FALSE</span>)</span></code></pre></div>
<div id="benchmarking-execution-time-1" class="section level3">
<h3>Benchmarking execution time</h3>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a>sample_id  <span class="ot">&lt;-</span> <span class="st">&quot;R.FileName&quot;</span> </span>
<span id="cb10-2"><a href="#cb10-2" tabindex="-1"></a></span>
<span id="cb10-3"><a href="#cb10-3" tabindex="-1"></a>secondary_id <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;EG.Library&quot;</span>, <span class="st">&quot;FG.Id&quot;</span>, <span class="st">&quot;FG.Charge&quot;</span>, <span class="st">&quot;F.FrgIon&quot;</span>, <span class="st">&quot;F.Charge&quot;</span>, <span class="st">&quot;F.FrgLossType&quot;</span>)</span>
<span id="cb10-4"><a href="#cb10-4" tabindex="-1"></a></span>
<span id="cb10-5"><a href="#cb10-5" tabindex="-1"></a>annotation_columns <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;PG.Genes&quot;</span>, <span class="st">&quot;PG.ProteinNames&quot;</span>)</span>
<span id="cb10-6"><a href="#cb10-6" tabindex="-1"></a></span>
<span id="cb10-7"><a href="#cb10-7" tabindex="-1"></a><span class="fu">system.time</span>({</span>
<span id="cb10-8"><a href="#cb10-8" tabindex="-1"></a>    </span>
<span id="cb10-9"><a href="#cb10-9" tabindex="-1"></a>    <span class="co"># reading data</span></span>
<span id="cb10-10"><a href="#cb10-10" tabindex="-1"></a>    raw <span class="ot">&lt;-</span> <span class="fu">read.delim</span>(<span class="st">&quot;DIA-report-long-format.txt&quot;</span>)</span>
<span id="cb10-11"><a href="#cb10-11" tabindex="-1"></a></span>
<span id="cb10-12"><a href="#cb10-12" tabindex="-1"></a>    <span class="co"># filtering</span></span>
<span id="cb10-13"><a href="#cb10-13" tabindex="-1"></a>    selected <span class="ot">&lt;-</span> raw<span class="sc">$</span>F.ExcludedFromQuantification <span class="sc">==</span> <span class="st">&quot;False&quot;</span> <span class="sc">&amp;</span> </span>
<span id="cb10-14"><a href="#cb10-14" tabindex="-1"></a>                <span class="sc">!</span><span class="fu">is.na</span>(raw<span class="sc">$</span>PG.Qvalue) <span class="sc">&amp;</span> raw<span class="sc">$</span>PG.Qvalue <span class="sc">&lt;</span> <span class="fl">0.01</span> <span class="sc">&amp;</span></span>
<span id="cb10-15"><a href="#cb10-15" tabindex="-1"></a>                <span class="sc">!</span><span class="fu">is.na</span>(raw<span class="sc">$</span>EG.Qvalue) <span class="sc">&amp;</span> raw<span class="sc">$</span>EG.Qvalue <span class="sc">&lt;</span> <span class="fl">0.01</span></span>
<span id="cb10-16"><a href="#cb10-16" tabindex="-1"></a></span>
<span id="cb10-17"><a href="#cb10-17" tabindex="-1"></a>    raw <span class="ot">&lt;-</span> raw[selected,]</span>
<span id="cb10-18"><a href="#cb10-18" tabindex="-1"></a></span>
<span id="cb10-19"><a href="#cb10-19" tabindex="-1"></a>    <span class="do">## process</span></span>
<span id="cb10-20"><a href="#cb10-20" tabindex="-1"></a></span>
<span id="cb10-21"><a href="#cb10-21" tabindex="-1"></a>    norm_data <span class="ot">&lt;-</span> iq<span class="sc">::</span><span class="fu">preprocess</span>(raw, </span>
<span id="cb10-22"><a href="#cb10-22" tabindex="-1"></a>                                <span class="at">sample_id  =</span> sample_id, </span>
<span id="cb10-23"><a href="#cb10-23" tabindex="-1"></a>                                <span class="at">secondary_id =</span> secondary_id)</span>
<span id="cb10-24"><a href="#cb10-24" tabindex="-1"></a></span>
<span id="cb10-25"><a href="#cb10-25" tabindex="-1"></a>    protein_list <span class="ot">&lt;-</span> iq<span class="sc">::</span><span class="fu">create_protein_list</span>(norm_data)</span>
<span id="cb10-26"><a href="#cb10-26" tabindex="-1"></a>    </span>
<span id="cb10-27"><a href="#cb10-27" tabindex="-1"></a>    result <span class="ot">&lt;-</span> iq<span class="sc">::</span><span class="fu">create_protein_table</span>(protein_list)</span>
<span id="cb10-28"><a href="#cb10-28" tabindex="-1"></a>    </span>
<span id="cb10-29"><a href="#cb10-29" tabindex="-1"></a>})</span>
<span id="cb10-30"><a href="#cb10-30" tabindex="-1"></a><span class="co">#&gt; Concatenating secondary ids...</span></span>
<span id="cb10-31"><a href="#cb10-31" tabindex="-1"></a><span class="co">#&gt; Removing low intensities...</span></span>
<span id="cb10-32"><a href="#cb10-32" tabindex="-1"></a><span class="co">#&gt; Barplotting raw data ...</span></span>
<span id="cb10-33"><a href="#cb10-33" tabindex="-1"></a><span class="co">#&gt; Median normalization ...</span></span>
<span id="cb10-34"><a href="#cb10-34" tabindex="-1"></a><span class="co">#&gt; Barplotting after normalization ...</span></span>
<span id="cb10-35"><a href="#cb10-35" tabindex="-1"></a><span class="co">#&gt; # proteins = 3554, # samples = 24</span></span>
<span id="cb10-36"><a href="#cb10-36" tabindex="-1"></a><span class="co">#&gt; 5%</span></span>
<span id="cb10-37"><a href="#cb10-37" tabindex="-1"></a><span class="co">#&gt; 10%</span></span>
<span id="cb10-38"><a href="#cb10-38" tabindex="-1"></a><span class="co">#&gt; 15%</span></span>
<span id="cb10-39"><a href="#cb10-39" tabindex="-1"></a><span class="co">#&gt; 20%</span></span>
<span id="cb10-40"><a href="#cb10-40" tabindex="-1"></a><span class="co">#&gt; 25%</span></span>
<span id="cb10-41"><a href="#cb10-41" tabindex="-1"></a><span class="co">#&gt; 30%</span></span>
<span id="cb10-42"><a href="#cb10-42" tabindex="-1"></a><span class="co">#&gt; 35%</span></span>
<span id="cb10-43"><a href="#cb10-43" tabindex="-1"></a><span class="co">#&gt; 40%</span></span>
<span id="cb10-44"><a href="#cb10-44" tabindex="-1"></a><span class="co">#&gt; 45%</span></span>
<span id="cb10-45"><a href="#cb10-45" tabindex="-1"></a><span class="co">#&gt; 50%</span></span>
<span id="cb10-46"><a href="#cb10-46" tabindex="-1"></a><span class="co">#&gt; 55%</span></span>
<span id="cb10-47"><a href="#cb10-47" tabindex="-1"></a><span class="co">#&gt; 60%</span></span>
<span id="cb10-48"><a href="#cb10-48" tabindex="-1"></a><span class="co">#&gt; 65%</span></span>
<span id="cb10-49"><a href="#cb10-49" tabindex="-1"></a><span class="co">#&gt; 70%</span></span>
<span id="cb10-50"><a href="#cb10-50" tabindex="-1"></a><span class="co">#&gt; 75%</span></span>
<span id="cb10-51"><a href="#cb10-51" tabindex="-1"></a><span class="co">#&gt; 80%</span></span>
<span id="cb10-52"><a href="#cb10-52" tabindex="-1"></a><span class="co">#&gt; 85%</span></span>
<span id="cb10-53"><a href="#cb10-53" tabindex="-1"></a><span class="co">#&gt; 90%</span></span>
<span id="cb10-54"><a href="#cb10-54" tabindex="-1"></a><span class="co">#&gt; 95%</span></span>
<span id="cb10-55"><a href="#cb10-55" tabindex="-1"></a><span class="co">#&gt; 100%</span></span>
<span id="cb10-56"><a href="#cb10-56" tabindex="-1"></a><span class="co">#&gt; Completed.</span></span>
<span id="cb10-57"><a href="#cb10-57" tabindex="-1"></a><span class="co">#&gt; 5%</span></span>
<span id="cb10-58"><a href="#cb10-58" tabindex="-1"></a><span class="co">#&gt; 10%</span></span>
<span id="cb10-59"><a href="#cb10-59" tabindex="-1"></a><span class="co">#&gt; 15%</span></span>
<span id="cb10-60"><a href="#cb10-60" tabindex="-1"></a><span class="co">#&gt; 20%</span></span>
<span id="cb10-61"><a href="#cb10-61" tabindex="-1"></a><span class="co">#&gt; 25%</span></span>
<span id="cb10-62"><a href="#cb10-62" tabindex="-1"></a><span class="co">#&gt; 30%</span></span>
<span id="cb10-63"><a href="#cb10-63" tabindex="-1"></a><span class="co">#&gt; 35%</span></span>
<span id="cb10-64"><a href="#cb10-64" tabindex="-1"></a><span class="co">#&gt; 40%</span></span>
<span id="cb10-65"><a href="#cb10-65" tabindex="-1"></a><span class="co">#&gt; 45%</span></span>
<span id="cb10-66"><a href="#cb10-66" tabindex="-1"></a><span class="co">#&gt; 50%</span></span>
<span id="cb10-67"><a href="#cb10-67" tabindex="-1"></a><span class="co">#&gt; 55%</span></span>
<span id="cb10-68"><a href="#cb10-68" tabindex="-1"></a><span class="co">#&gt; 60%</span></span>
<span id="cb10-69"><a href="#cb10-69" tabindex="-1"></a><span class="co">#&gt; 65%</span></span>
<span id="cb10-70"><a href="#cb10-70" tabindex="-1"></a><span class="co">#&gt; 70%</span></span>
<span id="cb10-71"><a href="#cb10-71" tabindex="-1"></a><span class="co">#&gt; 75%</span></span>
<span id="cb10-72"><a href="#cb10-72" tabindex="-1"></a><span class="co">#&gt; 80%</span></span>
<span id="cb10-73"><a href="#cb10-73" tabindex="-1"></a><span class="co">#&gt; 85%</span></span>
<span id="cb10-74"><a href="#cb10-74" tabindex="-1"></a><span class="co">#&gt; 90%</span></span>
<span id="cb10-75"><a href="#cb10-75" tabindex="-1"></a><span class="co">#&gt; 95%</span></span>
<span id="cb10-76"><a href="#cb10-76" tabindex="-1"></a><span class="co">#&gt; 100%</span></span>
<span id="cb10-77"><a href="#cb10-77" tabindex="-1"></a><span class="co">#&gt; Completed.</span></span>
<span id="cb10-78"><a href="#cb10-78" tabindex="-1"></a><span class="co">#&gt;    user  system elapsed </span></span>
<span id="cb10-79"><a href="#cb10-79" tabindex="-1"></a><span class="co">#&gt;  569.83   14.78  584.75</span></span>
<span id="cb10-80"><a href="#cb10-80" tabindex="-1"></a></span>
<span id="cb10-81"><a href="#cb10-81" tabindex="-1"></a><span class="fu">system.time</span>({</span>
<span id="cb10-82"><a href="#cb10-82" tabindex="-1"></a>    iq_dat <span class="ot">&lt;-</span> iq<span class="sc">::</span><span class="fu">fast_read</span>(<span class="st">&quot;DIA-report-long-format.txt&quot;</span>,</span>
<span id="cb10-83"><a href="#cb10-83" tabindex="-1"></a>                            <span class="at">sample_id  =</span> sample_id, </span>
<span id="cb10-84"><a href="#cb10-84" tabindex="-1"></a>                            <span class="at">secondary_id =</span> secondary_id,</span>
<span id="cb10-85"><a href="#cb10-85" tabindex="-1"></a>                            <span class="at">filter_string_equal =</span> <span class="fu">c</span>(<span class="st">&quot;F.ExcludedFromQuantification&quot;</span> <span class="ot">=</span> <span class="st">&quot;False&quot;</span>),</span>
<span id="cb10-86"><a href="#cb10-86" tabindex="-1"></a>                            <span class="at">annotation_col =</span> annotation_columns)</span>
<span id="cb10-87"><a href="#cb10-87" tabindex="-1"></a></span>
<span id="cb10-88"><a href="#cb10-88" tabindex="-1"></a>    iq_norm_data <span class="ot">&lt;-</span> iq<span class="sc">::</span><span class="fu">fast_preprocess</span>(iq_dat<span class="sc">$</span>quant_table)</span>
<span id="cb10-89"><a href="#cb10-89" tabindex="-1"></a></span>
<span id="cb10-90"><a href="#cb10-90" tabindex="-1"></a>    result_fastest <span class="ot">&lt;-</span> iq<span class="sc">::</span><span class="fu">fast_MaxLFQ</span>(iq_norm_data, </span>
<span id="cb10-91"><a href="#cb10-91" tabindex="-1"></a>                                      <span class="at">row_names =</span> iq_dat<span class="sc">$</span>protein[, <span class="dv">1</span>], </span>
<span id="cb10-92"><a href="#cb10-92" tabindex="-1"></a>                                      <span class="at">col_names =</span> iq_dat<span class="sc">$</span>sample)</span>
<span id="cb10-93"><a href="#cb10-93" tabindex="-1"></a>})</span>
<span id="cb10-94"><a href="#cb10-94" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb10-95"><a href="#cb10-95" tabindex="-1"></a><span class="co">#&gt; Command: --sample R.FileName --primary PG.ProteinGroups --secondary EG.Library FG.Id FG.Charge F.FrgIon F.Charge F.FrgLossType --quant F.PeakArea --annotation PG.Genes PG.ProteinNames --filter-string-equal F.ExcludedFromQuantification False --filter-double-less PG.Qvalue 0.01 --filter-double-less EG.Qvalue 0.01 DIA-report-long-format.txt </span></span>
<span id="cb10-96"><a href="#cb10-96" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb10-97"><a href="#cb10-97" tabindex="-1"></a><span class="co">#&gt; Sample column:</span></span>
<span id="cb10-98"><a href="#cb10-98" tabindex="-1"></a><span class="co">#&gt;     R.FileName</span></span>
<span id="cb10-99"><a href="#cb10-99" tabindex="-1"></a><span class="co">#&gt; Protein column:</span></span>
<span id="cb10-100"><a href="#cb10-100" tabindex="-1"></a><span class="co">#&gt;     PG.ProteinGroups</span></span>
<span id="cb10-101"><a href="#cb10-101" tabindex="-1"></a><span class="co">#&gt; Ion column(s):</span></span>
<span id="cb10-102"><a href="#cb10-102" tabindex="-1"></a><span class="co">#&gt;     EG.Library FG.Id FG.Charge F.FrgIon F.Charge F.FrgLossType</span></span>
<span id="cb10-103"><a href="#cb10-103" tabindex="-1"></a><span class="co">#&gt; Quant column:</span></span>
<span id="cb10-104"><a href="#cb10-104" tabindex="-1"></a><span class="co">#&gt;     F.PeakArea</span></span>
<span id="cb10-105"><a href="#cb10-105" tabindex="-1"></a><span class="co">#&gt; Annotation column(s):</span></span>
<span id="cb10-106"><a href="#cb10-106" tabindex="-1"></a><span class="co">#&gt;     PG.Genes PG.ProteinNames</span></span>
<span id="cb10-107"><a href="#cb10-107" tabindex="-1"></a><span class="co">#&gt; String equal filter(s):</span></span>
<span id="cb10-108"><a href="#cb10-108" tabindex="-1"></a><span class="co">#&gt;     F.ExcludedFromQuantification == False</span></span>
<span id="cb10-109"><a href="#cb10-109" tabindex="-1"></a><span class="co">#&gt; Double less filter(s):</span></span>
<span id="cb10-110"><a href="#cb10-110" tabindex="-1"></a><span class="co">#&gt;     PG.Qvalue &lt; 0.010000</span></span>
<span id="cb10-111"><a href="#cb10-111" tabindex="-1"></a><span class="co">#&gt;     EG.Qvalue &lt; 0.010000</span></span>
<span id="cb10-112"><a href="#cb10-112" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb10-113"><a href="#cb10-113" tabindex="-1"></a><span class="co">#&gt; Using 4 threads ...</span></span>
<span id="cb10-114"><a href="#cb10-114" tabindex="-1"></a><span class="co">#&gt; 20 samples read</span></span>
<span id="cb10-115"><a href="#cb10-115" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb10-116"><a href="#cb10-116" tabindex="-1"></a><span class="co">#&gt; # lines read (excluding headers)      = 5547331</span></span>
<span id="cb10-117"><a href="#cb10-117" tabindex="-1"></a><span class="co">#&gt; # quantitative values after filtering = 3390569</span></span>
<span id="cb10-118"><a href="#cb10-118" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb10-119"><a href="#cb10-119" tabindex="-1"></a><span class="co">#&gt; # samples  = 24</span></span>
<span id="cb10-120"><a href="#cb10-120" tabindex="-1"></a><span class="co">#&gt; # proteins = 3554</span></span>
<span id="cb10-121"><a href="#cb10-121" tabindex="-1"></a><span class="co">#&gt; Removing low intensities...</span></span>
<span id="cb10-122"><a href="#cb10-122" tabindex="-1"></a><span class="co">#&gt; Barplotting raw data ...</span></span>
<span id="cb10-123"><a href="#cb10-123" tabindex="-1"></a><span class="co">#&gt; Median normalization ...</span></span>
<span id="cb10-124"><a href="#cb10-124" tabindex="-1"></a><span class="co">#&gt; Barplotting after normalization ...</span></span>
<span id="cb10-125"><a href="#cb10-125" tabindex="-1"></a><span class="co">#&gt; nrow = 3369557, # proteins = 3554, # samples = 24</span></span>
<span id="cb10-126"><a href="#cb10-126" tabindex="-1"></a><span class="co">#&gt; Using 35 threads...</span></span>
<span id="cb10-127"><a href="#cb10-127" tabindex="-1"></a><span class="co">#&gt; 0%</span></span>
<span id="cb10-128"><a href="#cb10-128" tabindex="-1"></a><span class="co">#&gt; 5%</span></span>
<span id="cb10-129"><a href="#cb10-129" tabindex="-1"></a><span class="co">#&gt; 13%</span></span>
<span id="cb10-130"><a href="#cb10-130" tabindex="-1"></a><span class="co">#&gt; 18%</span></span>
<span id="cb10-131"><a href="#cb10-131" tabindex="-1"></a><span class="co">#&gt; 24%</span></span>
<span id="cb10-132"><a href="#cb10-132" tabindex="-1"></a><span class="co">#&gt; 29%</span></span>
<span id="cb10-133"><a href="#cb10-133" tabindex="-1"></a><span class="co">#&gt; 34%</span></span>
<span id="cb10-134"><a href="#cb10-134" tabindex="-1"></a><span class="co">#&gt; 40%</span></span>
<span id="cb10-135"><a href="#cb10-135" tabindex="-1"></a><span class="co">#&gt; 45%</span></span>
<span id="cb10-136"><a href="#cb10-136" tabindex="-1"></a><span class="co">#&gt; 50%</span></span>
<span id="cb10-137"><a href="#cb10-137" tabindex="-1"></a><span class="co">#&gt; 55%</span></span>
<span id="cb10-138"><a href="#cb10-138" tabindex="-1"></a><span class="co">#&gt; 61%</span></span>
<span id="cb10-139"><a href="#cb10-139" tabindex="-1"></a><span class="co">#&gt; 68%</span></span>
<span id="cb10-140"><a href="#cb10-140" tabindex="-1"></a><span class="co">#&gt; 75%</span></span>
<span id="cb10-141"><a href="#cb10-141" tabindex="-1"></a><span class="co">#&gt; 80%</span></span>
<span id="cb10-142"><a href="#cb10-142" tabindex="-1"></a><span class="co">#&gt; 86%</span></span>
<span id="cb10-143"><a href="#cb10-143" tabindex="-1"></a><span class="co">#&gt; 92%</span></span>
<span id="cb10-144"><a href="#cb10-144" tabindex="-1"></a><span class="co">#&gt; 98%</span></span>
<span id="cb10-145"><a href="#cb10-145" tabindex="-1"></a><span class="co">#&gt; Completed.</span></span>
<span id="cb10-146"><a href="#cb10-146" tabindex="-1"></a><span class="co">#&gt;    user  system elapsed </span></span>
<span id="cb10-147"><a href="#cb10-147" tabindex="-1"></a><span class="co">#&gt;   27.54    1.56   12.34</span></span></code></pre></div>
</div>
</div>
<div id="references" class="section level2">
<h2>References</h2>
<ol style="list-style-type: decimal">
<li>Pham TV, Henneman AA, Jimenez CR (2020) iq: an R package to estimate
relative protein abundances from ion quantification in DIA-MS-based
proteomics. <em>Bioinformatics</em> 36(8):2611-2613.</li>
</ol>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
